langserver/loader.go

```diff
@@ -306,3 +312,47 @@ func isMultiplePackageError(err error) bool {
 	_, ok := err.(*build.MultiplePackageError)
 	return ok
 }
+
+var freeOSMemoryOnce sync.Once
+
+// freeOSMemory spawns a goroutine (only once) which invokes
+// runtime/debug.FreeOSMemory() more aggressively than the runtime default of
+// 5 minutes after GC.
+//
+// There is a long-standing known issue with Go in which memory is not returned
+// to the OS aggressively enough[1], which coincidently harms our application
+// quite a lot because we perform so many short-burst heap allocations during
+// the type-checking phase.
+//
+// This function should only be invoked in editor mode, not in sourcegraph.com
+// mode, because users running the language server as part of their editor
+// generally expect much lower memory usage. In contrast, on sourcegraph.com we
+// can give our servers plenty of RAM and allow Go to consume as much as it
+// wants. Go does reuse the memory not free'd to the OS, and as such enabling
+// this does _technically_ make our application perform less optimally -- but
+// in practice this has no observable affect in editor mode.
+//
+// The end effect of performing this is that repeating "hover over code" -> "make an edit"
+// 10 times inside a large package like github.com/docker/docker/cmd/dockerd:
+//
+//
+// 	| Real Before | Real After | Real Change | Go Before | Go After | Go Change |
+// 	|-------------|------------|-------------|-----------|----------|-----------|
+// 	| 7.61GB      | 4.12GB     | -45.86%     | 3.92GB    | 3.33GB   | -15.05%   |
```

Comment:
Great question.

I have observed that both before this change and after this change, the memory does plateau at a certain amount (8.3GB before this change, and 7.22GB after this change) and does not continue to grow. That amount can just be quite large for projects with many dependencies.

The amount appears to be 2x the currently in-use Go heap size (~3.3GB), which makes sense. It's not clear to me why Go can only release 1GB of the ~4 GB difference in unused memory, but I speculate it could be due to issues like https://github.com/golang/go/issues/14045 or https://github.com/golang/go/issues/14735 since our allocations all come in large spikes due to type-checking.

If we can make type-checking faster more fine-grain caching (which I am doing), then we can potentially spare ourselves keeping a ~3.3GB Go heap in-memory all the time and lower this number even further (in addition to performing less type-checking on-the-fly).